apps/api/src/deployment/services/top-up-deployments/top-up-deployments.service.ts

```diff
@@ -1,12 +1,135 @@
+import { AllowanceHttpService, BalanceHttpService, DeploymentAllowance } from "@akashnetwork/http-sdk";
+import { PromisePool } from "@supercharge/promise-pool";
 import { singleton } from "tsyringe";
 
+import { InjectWallet } from "@src/billing/providers/wallet.provider";
+import { UserWalletOutput, UserWalletRepository } from "@src/billing/repositories";
+import { MasterWalletService } from "@src/billing/services";
 import { LoggerService } from "@src/core";
+import { DrainingDeploymentOutput, LeaseRepository } from "@src/deployment/repositories/lease/lease.repository";
+
+interface Balances {
+  denom: string;
+  feesLimit: number;
+  deploymentLimit: number;
+  balance: number;
+  isManaged: boolean;
+}
 
 @singleton()
 export class TopUpDeploymentsService {
+  private readonly CONCURRENCY = 10;
+
   private readonly logger = new LoggerService({ context: TopUpDeploymentsService.name });
 
+  constructor(
+    private readonly userWalletRepository: UserWalletRepository,
+    private readonly allowanceHttpService: AllowanceHttpService,
+    private readonly balanceHttpService: BalanceHttpService,
+    @InjectWallet("MANAGED") private readonly managedMasterWalletService: MasterWalletService,
+    @InjectWallet("UAKT_TOP_UP") private readonly uaktMasterWalletService: MasterWalletService,
+    @InjectWallet("USDC_TOP_UP") private readonly usdtMasterWalletService: MasterWalletService,
+    private readonly leaseRepository: LeaseRepository
+  ) {}
+
   async topUpDeployments() {
-    this.logger.warn("Top up deployments not implemented");
+    const wallets = [this.uaktMasterWalletService, this.usdtMasterWalletService];
+
+    const topUpAllManagedDeployments = wallets.map(async wallet => {
+      const address = await wallet.getFirstAddress();
+      await this.allowanceHttpService.paginateDeploymentGrantsForGrantee(address, async grants => {
+        await PromisePool.withConcurrency(this.CONCURRENCY)
+          .for(grants)
+          .process(async grant => this.topUpForGrant(grant));
+      });
+    });
+    await Promise.all(topUpAllManagedDeployments);
+
+    await this.paginateManagedWallets(async userWallets => {
+      await Promise.all(userWallets.map(async userWallet => this.topUpForManagedWallet(userWallet)));
+    });
+  }
+
+  private async topUpForGrant(grant: DeploymentAllowance) {
+    const balances = await this.collectCustodialWalletBalances(grant);
+    const owner = grant.granter;
+    this.logger.debug({ event: "BALANCES_COLLECTED", granter: owner, grantee: grant.grantee, balances });
+
+    const drainingDeployments = await this.retrieveDrainingDeployments(owner);
+
+    drainingDeployments.map(async deployment => {
+      const topUpAmount = await this.calculateTopUpAmount(deployment);
+      this.validateTopUpAmount(topUpAmount, balances);
+    });
+  }
+
+  private async collectCustodialWalletBalances(grant: DeploymentAllowance): Promise<Balances> {
```

Comment:
this is so far is the only functional method here along with the ones calling it